src/analytics-preload.ts

```diff
@@ -0,0 +1,271 @@
+
+/**
+ *  Represents a buffered method call that occurred before initialization.
+ */
+export interface PreInitMethodCall<
+  T extends PreInitMethodName = PreInitMethodName
+> {
+  method: T
+  args: Parameters<Analytics[T]>
+  called: boolean
+  resolve: (v: ReturnTypeUnwrap<Analytics[T]>) => void
+}
+
+const normalizeSnippetBuffer = (buffer: SnippetBuffer): PreInitMethodCall[] => {
+  return buffer.map(
+    ([methodName, ...args]) =>
+      ({
+        method: methodName,
+        resolve: () => {},
+        args,
+        called: false,
+      } as PreInitMethodCall)
+  )
+}
+
+type SnippetWindowBufferedMethodArgs = any[]
+
+type SnippetWindowBufferedMethodCall = [
+  PreInitMethodName,
+  ...SnippetWindowBufferedMethodArgs
+]
+
+/**
+ * A list of the method calls before initialization for snippet users
+ * For example, [["track", "foo", {bar: 123}], ["page"], ["on", "ready", function(){..}]
+ */
+type SnippetBuffer = SnippetWindowBufferedMethodCall[]
+
+/**
+ * Fetch the buffered method calls from the window object and normalize them.
+ */
+const getSnippetWindowBuffer = (): PreInitMethodCall[] => {
+  const wa = window.analytics
+  const buffered =
+    // @ts-expect-error
+    (wa && wa[0] ? [...wa] : []) as SnippetBuffer
+  return normalizeSnippetBuffer(buffered)
+}
+/**
+ * Infer return type; if return type is promise, unwrap it.
+ */
+type ReturnTypeUnwrap<Fn> = Fn extends (...args: any[]) => infer ReturnT
+  ? ReturnT extends PromiseLike<infer Unwrapped>
+    ? Unwrapped
+    : ReturnT
+  : never
+
+export class PreInitMethodCallBuffer {
+  /**
+   *  Represents any and all the buffered method calls that occurred before initialization.
+   */
+  private _list: PreInitMethodCall[] = []
+
+  private _windowBuffer: PreInitMethodCall[] = []
+
+  public saveSnippetWindowBuffer() {
+    // store reference in order to be be to mutate 'method.called' field from callers.
+    this._windowBuffer = getSnippetWindowBuffer()
+  }
+
+  public get list(): PreInitMethodCall[] {
+    return [...this._list, ...this._windowBuffer]
+  }
+
+  push(...calls: PreInitMethodCall[]) {
+    this._list.push(...calls)
+  }
+
+  clear(): void {
+    this._list = []
+  }
+
+  /**
+   *  Call method and mark as "called"
+   */
+  async callMethod<T extends PreInitMethodName>(
+    analytics: Analytics,
+    methodCall: PreInitMethodCall<T>
+  ) {
+    if (methodCall.called) {
+      return undefined
+    }
+
+    methodCall.called = true
+
+    const { method, args, resolve } = methodCall
+    const result = await (analytics[method] as Function)(...args)
+    resolve(result)
+  }
+}
+
+/**
+ * usage:
+ * const analytics = new AnalyticsInitBuffer(AnalyticsBrowser.load(...))
+ *
+ * analytics.track("foo")
+ *
+ * const analyticsRealInstance = await analytics
+ *
+ */
+
+/**
+ * Inject a preInitBuffer into the getAnalytics function so load can call
+ */
+type Executor = (
+  preInitBuffer: PreInitMethodCallBuffer
+) => Promise<[Analytics, Context]>
+
```

Comment:
WIP -- mulling over having this whole class have a method that returns an extended Promise (can't extend it directly, per Liskov Substitution Principle).
e.g.
```ts
const thenable = new AnalyticsWithBuffer((preInitBuffer) => AnalyticsBrowser.load(arg1, preInitBuffer))
  .createThenableInstance()
```